packages/tasit-identity-contract/src/GnosisSafe.test.js

```diff
@@ -0,0 +1,152 @@
+// in this test suite. Eventually, maybe ProviderFactory may move to
+// some shared helper dir.
+import ProviderFactory from "tasit-action/dist/ProviderFactory";
+
+const { GnosisSafe: GNOSIS_SAFE_ADDRESS } = localAddresses;
+const ERC20_ADDRESS = "0x37E1A58dD465D33263D00185D065Ee36DD34CDb4";
+
+const { utils: ethersUtils } = ethers;
+const { bigNumberify } = ethersUtils;
+
+const ZERO = 0;
+const ONE = bigNumberify(`${1e18}`);
+
+// 100 gwei
+const GAS_PRICE = bigNumberify(`${1e11}`);
+
+describe("GnosisSafe", () => {
+  const config = {
+    provider: {
+      network: "other",
+      provider: "jsonrpc",
+      pollingInterval: 50,
+      jsonRpc: {
+        url: "http://localhost",
+        port: 8545,
+      },
+    },
+    events: {
+      timeout: 2000,
+    },
+  };
+  let gnosisSafeContract;
+  let gnosisSafeUtils;
+  let anaWallet;
+  let snapshotId;
+  let provider;
+  let erc20;
+
+  before("", async () => {
+    ConfigLoader.setConfig(config);
+
+    provider = ProviderFactory.getProvider();
+
+    const anaPrivateKey =
+      "0x11d943d7649fbdeb146dc57bd9cfc80b086bfab2330c7b25651dbaf382392f60";
+
+    anaWallet = createFromPrivateKey(anaPrivateKey);
+
+    // Contract deployment setup with ana (accounts[0]) as the only owner
+    // To change that, edit the file "tasit-contract/3rd-parties/gnosis/scripts/2_deploy_contracts.js"
+    gnosisSafeContract = new GnosisSafe(GNOSIS_SAFE_ADDRESS);
+
+    erc20 = new DetailedERC20(ERC20_ADDRESS);
+  });
+
+  beforeEach("", async () => {
+    snapshotId = await createSnapshot(provider);
+
+    const etherBalance = await provider.getBalance(GNOSIS_SAFE_ADDRESS);
+    expect(`${etherBalance}`).to.equal(`${ZERO}`);
+
+    const erc20Balance = await erc20.balanceOf(GNOSIS_SAFE_ADDRESS);
+    expect(`${erc20Balance}`).to.equal(`${ZERO}`);
+
+    //gnosisSafeContract.removeWallet();
+    //erc20.removeWallet();
+  });
+
+  afterEach("", async () => {
+    await revertFromSnapshot(provider, snapshotId);
+  });
+
+  it("wallet owner should deposit and withdraw some ethers", async () => {
+    anaWallet = anaWallet.connect(provider);
+    const tx = await anaWallet.sendTransaction({
+      to: GNOSIS_SAFE_ADDRESS,
+      value: ONE,
+    });
+    await provider.waitForTransaction(tx.hash);
+
+    const balanceAfterDeposit = await provider.getBalance(GNOSIS_SAFE_ADDRESS);
+    expect(`${balanceAfterDeposit}`).to.equal(`${ONE}`);
+
+    const signers = [anaWallet];
+    const toAddress = anaWallet.address;
+    const value = ONE;
+
+    gnosisSafeContract.setWallet(anaWallet);
+    const execTxAction = await gnosisSafeContract.sendEtherTransaction(
+      signers,
+      toAddress,
+      value
+    );
+    await execTxAction.waitForNonceToUpdate();
+
+    const balanceAfterWithdraw = await provider.getBalance(GNOSIS_SAFE_ADDRESS);
+    expect(`${balanceAfterWithdraw}`).to.equal(`${ZERO}`);
+  });
+
+  it("wallet owner should deposit and withdraw some ERC20 tokens", async () => {
```

Comment:
This test doesn't need `anaWallet = anaWallet.connect(provider);`?

Reply:
No, not really since the contract is already connected with the provider.

Reply:
But how is it connected with the provider? From the previous `it` test? I'd like to completely isolate the tests if possible.

Reply:
The config is load to the singleton `ConfigLoader` object when app starts. Always that a new contract is created, the provider is created too using `ProviderFactory` that uses config stored inside of `ConfigLoader`.

Reply:
Thanks. But then why does the preceding test case have `anaWallet = anaWallet.connect(provider);`?

Reply:
Because that was sent ethers directly from the wallet without using a connected contract. So, to broadcast a tx the wallet needs to be connected.

Reply:
Ah ok, whereas ERC20 must involve a connected contract. Got it.

Reply:
Resolving this.